app/services/activities/import_from_csv.rb

```diff
@@ -122,11 +122,10 @@ def update
           end
         end
 
-        return if @activity.save
+        return set_errors unless @activity.valid?
 
-        @activity.errors.each do |error|
-          @errors[error.attribute] ||= [@converter.raw(error.attribute), error.message]
-        end
+        record_history
```

Comment:
Do you think there is any value in recording the history after the save, incase anyting goes wrong during the save and we write a untrue history?

Reply:
yes, good point!

I think this behaviour is rooted back in the earlier Wizard implementation where there's no clear way in `Staff::ActivityFormsController` to create the History after the Activity has been updated:

https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/master/app/controllers/staff/activity_forms_controller.rb#L74-L79

as the persistence is performed magically by the Wizard gem.

But in the case of the bulk importer (`Activities::ImportFromCSV::ActivityUpdater`) we **can** handle the creation of history after the activity is updated.

It's worth noting that the 'main'  `Activities::ImportFromCSV#import` method wraps the row-at-a-time import processes in an `ActiveRecord::Base.transaction`, meaning that if there's any exception, then all changes (including the creation of `HistoricalEvent`s) is rolled back.

https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/master/app/services/activities/import_from_csv.rb#L32-L42

But yes what you suggest is definitely better -- clearer and safer. I've added a commit to add this behaviour.